src/components/ReadingProgressPage/ReadingHistory/AddReading/index.tsx

```diff
@@ -0,0 +1,287 @@
+  dateNumbersToString,
+  dateToReadableFormat,
+  dateToYearMonthDay,
+  getCurrentMonth,
+  getMonthDateObject,
+  getMonthsInYear,
+} from '@/utils/datetime';
+import { logValueChange, logButtonClick, logFormSubmission } from '@/utils/eventLogger';
+
+const AddReading = () => {
+  const chaptersData = useContext(DataContext);
+  const { t, lang } = useTranslation('reading-progress');
+  const [isOpen, setIsOpen] = useState(false);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const [rangeStart, setRangeStart] = useState<string | null>(null);
+  const [rangeEnd, setRangeEnd] = useState<string | null>(null);
+  const [ranges, setRanges] = useState<string[]>([]);
+  const [selectedMonth, setSelectedMonth] = useState<number>(() => getCurrentMonth());
+  const [selectedDate, setSelectedDate] = useState<string | null>(null);
+  const selectedYear = useMemo(() => new Date().getFullYear(), []);
+  const months = useMemo(() => getMonthsInYear(selectedYear, lang), [selectedYear, lang]);
+  const mushaf = useGetMushaf();
+  const toast = useToast();
+  const { cache, mutate } = useSWRConfig();
+
+  const onClose = () => {
+    setIsOpen(false);
+    // reset ranges
+    setRanges([]);
+    // reset selected date
+    setSelectedDate(null);
+  };
+
+  const onOpenClick = () => {
+    logButtonClick('open_add_reading_modal');
+    setIsOpen(true);
+  };
+
+  const selectedMonthObj = useMemo(() => {
+    if (!selectedMonth) return null;
+    return months.find((month) => month.id === selectedMonth);
+  }, [selectedMonth, months]);
+
+  const onMonthBackClick = () => {
+    const newMonth = selectedMonth - 1;
+    logValueChange('add_reading_month', selectedMonth, newMonth, {
+      year: selectedYear,
+    });
+    setSelectedMonth(newMonth);
+  };
+
+  const onMonthForwardClick = () => {
+    const newMonth = selectedMonth + 1;
+    logValueChange('add_reading_month', selectedMonth, newMonth, {
+      year: selectedYear,
+    });
+    setSelectedMonth(newMonth);
+  };
+
+  const onAddClick = () => {
+    if (!rangeStart || !rangeEnd) return;
+    const newRanges = [...ranges, `${rangeStart}-${rangeEnd}`];
+    logButtonClick('add_reading', {
+      range: `${rangeStart}-${rangeEnd}`,
+    });
+    setRanges(newRanges);
+    setRangeStart(undefined);
+    setRangeEnd(undefined);
+  };
+
+  const onGoBackClick = () => {
+    logButtonClick('add_reading_back_to_calendar');
+    setSelectedDate(null);
+  };
+
+  const onDayClick = (day: number, dateString: string) => {
+    logValueChange('add_reading_day', selectedDate, dateString, {
+      day,
+    });
+    setSelectedDate(dateString);
+  };
+
+  const onRangeChange = ({ startVerse, endVerse }) => {
+    setRangeStart(startVerse);
+    setRangeEnd(endVerse);
+  };
+
+  const getIsAddButtonDisabled = () => {
+    return !areValidReadingRanges(chaptersData, {
+      startVerse: rangeStart,
+      endVerse: rangeEnd,
+    });
+  };
+
+  const currentMonthDisabledDaysMap = useMemo(() => {
```

Comment:
nit: since we only need the map key (in this case the date string), we can refactor this to a `Set`